### PR TITLE
Correctly set out-of-boundary values.

### PR DESCRIPTION
**Pull request by eli-osherovich**
This is a fix for incorrectly set out-of-boundary pixels: #46637

### CODE REVIEW

**Review by mihaimaruseac — COMMENTED** [review]
Looks good, but can you also add a test please?

**Comment by gbaned** [comment]
@eli-osherovich  Can you please check @mihaimaruseac's comments and keep us posted ? Thanks!

**Comment by eli-osherovich** [comment]
I will add some tests shortly

**Comment by eli-osherovich** [comment]
@mihaimaruseac 
Done.

**Review by mihaimaruseac — APPROVED** [review]

**Review by hyeygit — COMMENTED** [review]
The issue reported in #46637 appears to be working-as-intended to me.

My understanding is that if you rotate
```
[[1, 1, 1],
 [1, 1, 1], 
 [1, 1, 1]]
```
by 45 degrees, then the _uncropped_ output should look like (with `fill_value=0`):
```
[0.        , 0.        , 0.3431458 , 0.        , 0.        ]
[0.        , 0.58578634, 1.        , 0.58578634, 0.        ]
[0.3431458 , 1.        , 1.        , 1.        , 0.3431458 ]
[0.        , 0.58578634, 1.        , 0.58578634, 0.        ]
[0.        , 0.        , 0.3431458 , 0.        , 0.        ]
```
But because the `output_shape` is set to (3, 3), we are only seeing the middle 3x3 patch. That is,`0.58578634` is actually within the image boundary and, therefore, should not be set to `fill_value`. The value is not `1` because the sides of the rotated square land between the pixels.

**Comment by eli-osherovich** [comment]
@hyeygit 
This is not how it works in all other packages (at least those that I am aware of including `keras-preprocessing`, which uses `scipy`). The flow that you describe is: Pad->Transform->Crop.  However, it is more common to not transform pixels that lie outside the original image boundaries but to use `fill_value` for those.

**Comment by mihaimaruseac** [comment]
The behavior described on https://github.com/tensorflow/tensorflow/pull/46752#pullrequestreview-589180100 is what happens when you rotate the image in Gimp too (for example).

I think we have so many tests that now rely on this behavior, so it is not that easy to change. This might be a reason why the internal tests failed.

**Comment by hyeygit** [comment]
@mihaimaruseac 
Thank you for looking into it.

@eli-osherovich 
Changing the existing behavior is not backwards compatible and may cause large breakage. We could add an option for it as an alternative.

For example, if we look at [`rotate`](https://scikit-image.org/docs/dev/api/skimage.transform.html#rotate) from `skimage`, there is a `clip` option that users can set.

_**clip** bool, optional
Whether to clip the output to the range of values of the input image. This is enabled by default, since higher order interpolation may produce values outside the given input range._

If we input 
```
a = [[1, 1, 1],
     [1, 1, 1], 
     [1, 1, 1]]
```
and set `clip=False`, we get the same behavior as `ImageProjectiveTransformV3` (#46637):
```
# `order=1` for 'BILINEAR'
skimage.transform.rotate(image=a, angle=45, order=1, clip=False)

array([[0.58578644, 1.        , 0.58578644],
       [1.        , 1.        , 1.        ],
       [0.58578644, 1.        , 0.58578644]])
```
`skimage` sets `clip` to `True` by default. Doing so clips the values outside the given input range: 
```
# `order=0` for 'NEAREST'
skimage.transform.rotate(image=a, angle=45, order=0, clip=True)

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])
```
```
# `order=1` for 'BILINEAR'
skimage.transform.rotate(image=a, angle=45, order=1, clip=True)
array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])
```
```
# `order=2` for 'BIQUADRATIC`
skimage.transform.rotate(image=a, angle=45, order=2, clip=True)
array([[0., 1., 0.],
       [1., 1., 1.],
       [0., 1., 0.]])
```

Notice that `NEAREST` and `BILINEAR` results are identical. We see the effect of `clip`ping with higher order interpolations which [`ImageProjectiveTransformV3`](https://www.tensorflow.org/api_docs/python/tf/raw_ops/ImageProjectiveTransformV3) currently doesn't support.

**Comment by eli-osherovich** [comment]
@hyeygit 
Thanks for your prompt response. I think the best counterpart in `skimage` is `skimage.transform.warp` because `rotate` does additional things (padding before transforming). 

At any rate, I personally, do not see a big problem here. Even though I would really prefer consistent behavior across different packages in the same ecosystem/software stack.

**Comment by eli-osherovich** [comment]
I have to correct myself:  `rotate` does not pad image by default. It calls `warp` pretty much straight forward.

**Comment by eli-osherovich** [comment]
In fact, `skimage` looks buggy to me after I dig into the internals. 
In the code below, `scipy` produces the result that I expect, whereas `skimage` does something wrong:
```python
a = np.ones((3, 3))
transform =  skimage.transform.SimilarityTransform(translation=(-1,-1)) + skimage.transform.SimilarityTransform(rotation=np.deg2rad(45)) + skimage.transform.SimilarityTransform(translation=(1,1))

skimage.transform.warp(a, transform, order=1, cval=42)
array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

scipy.ndimage.affine_transform(a, transform.params, order=1, cval=42)
array([[42.,  1., 42.],
       [ 1.,  1.,  1.],
       [42.,  1., 42.]])
```

**Comment by gbaned** [comment]
@hyeygit  Can you please take a look on above comments from @eli-osherovich. Thanks!

**Comment by eli-osherovich** [comment]
What is the verdict here? Will it stay as it is today?

**Comment by hyeygit** [comment]
Sorry for the late reply. 

@eli-osherovich 
Thank you for providing additional explanation and examples.
Also, I agree that having consistent behavior across different packages is easier for users.

Regarding merging this PR, it should not be a problem as long as there aren't things that rely on current behavior and all checks pass. I'll bring in the PR and run internal tests.

**Comment by hyeygit** [comment]
Unfortunately, there are many internal targets that rely on the current behavior and this V3 op would need to stay as it is. (Introducing a V4 op with a flag for switching between the two behavior is an option but would need API owner's approval.)

**Comment by gbaned** [comment]
@hyeygit  Any update on this PR? Please. Thanks!

**Comment by hyeygit** [comment]
@eli-osherovich We unfortunately cannot merge in the PR due to failing internal targets that rely on the default behavior which is WAI. May we close the PR?

**Comment by eli-osherovich** [comment]
Sure.
